cmake/clang-tidy.cmake

```diff
@@ -0,0 +1,31 @@
+find_program(CLANG_TIDY clang-tidy)
+if(NOT CLANG_TIDY)
+    message(STATUS "Did not find clang-tidy, target tidy is disabled.")
+else()
+    message(STATUS "Found clang-tidy, use \"make tidy\" to run it.")
+
+    set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+
+    set(CLANG_TIDY_CHECKS "*")
+    set(CLANG_TIDY_CHECKS "${CLANG_TIDY_CHECKS},-clang-analyzer-*")
+    set(CLANG_TIDY_CHECKS "${CLANG_TIDY_CHECKS},-clang-analyzer-cplusplus*")
+    set(CLANG_TIDY_CHECKS "-checks='${CLANG_TIDY_CHECKS}'")
+
+    macro(clang_tidy directory DESC)
```

Comment:
What's the reason for these macros?

Reply:
One is to list all the directories s and second one is to recursively search the directories and add the  `*.cpp` and `*.hpp` files to apply clang-tidy to.

Reply:
I seem to be missing something. Why can't we use the compile-database directly?

Reply:
We can us it using this [script](https://github.com/llvm-mirror/clang-tools-extra/blob/master/clang-tidy/tool/run-clang-tidy.py). From the man pages I understood that it needs the paths to the file which it then compares with in the database, correct me if understood incorrectly.

Reply:
If we're using that script, technically we don't need to supply the list of files (or the database, since it does auto discovery. See code). My knowledge and usage is based on [this](https://www.kdab.com/clang-tidy-part-1-modernize-source-code-using-c11c14).

Regarding the files, do you see a reason for
* providing all the file names: Possibly in future, it maybe used to exclude 3dparty lib but you're not doing that
* doing it per-directory instead of system: I assumed so far all fixes have been for the complete library.

Reply:
Both the reasons you mentioned can be integrated into the script itself so using the script offers everything we can do with CMake + extra stuff like auto-discovery and other features. 

Will go integrate instead then if you don't see any potential issues in using the script for all 
of this.

Reply:
IMO, using the script is nice, and as such, we can offload some CMake to that script instead. With that in mind, I don't see a reason for the 2 things you're doing that require CMake, namely
* passing list of files (and cmd db)
* doing it per directory
I was what rational was there behind your choice for adding the CMake

Reply:
Well now I had time to do a short check:
- Use `run-clang-tidy` should be the choice, as it can run multithreaded, can installed via apt-get and is less code we have to maintain
- `run-clang-tidy` allows export results as aggregated is as YAML - in case it makes it easier to integrate it into CI (we are using a Jenkins plugin, thats why we don't need it in our project)
- Filtering the result is the ugly job. As long we don't have this I don't see a big advantage by integrate it into this project - ant not sure if you want to do this really via CMake - even it may be possible ;-)

On our CI we are using it via:
```
cmake .. -DCMAKE_BUILD_TYPE=Debug -DCMAKE_GLOBAL_AUTOGEN_TARGET=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=ON ...
run-clang-tidy-10 -header-filter='^((src)|(tests))/.*\$' -checks='-*,bugprone-*,-bugprone-narrowing-conversions,cert-*,...,performance-*' > clangtidy_result.txt
```
Of course instead of passing checks via `-checks` we should better use a `.clang-format`-file

Reply:
Thanks for looking into it. If that's the case, we can make a config file for the 2 inputs:
* header filter
* checks

The config file allow us to version control these separately from the glue-code. We also have 2 usecases and so we need 2 entry points:
* Use a bash script so we can run this on format CI without CMake
* CMake to allow `make tidy` for the contributors (preferably using the same script to reduce duplication)

Reply:
Header filter and checks can both be defined in `.clang-tidy` . Just use `--dump-config` to create the file from my command line.

Just care: You can only filter header files - not source files! So we still need a manual filter.

Reply:
Apologies in advance for this tangential question: what's the reason for filtering out source files? Do they get picked up anyways using the compilation db or do they cause unnecessary issues for clang-tidy?

Reply:
In general you have two cases:
- You don't want that clang-tidy check 3rd-Party code, as clang-tidy can be time expensive and we don't want adjust 3rd-Party code, so we don't need the warning there
- You want disable checks for some directories (GoogleTest is a pain with clang-tidy as the macros will expand to code clang-tidy likes to put to the pillory).

Reply:
Just to summarize everything so I am on the right track:
1. Use run-clang-tidy
2. Read the checks and header filters from from `.clang-tidy`
3. Create bash script to run on CI and a CMake version for people to use

> So we still need a manual filter
How do you recommend to go on doing this, since we are reading the list of file from `compile_commands.json`?

Reply:
Point 1 & 3: 👍 
Point 2 is unnecessary, as clang-tidy will already do it ;-)

@filter: I don't know , can you check if in the YAML files are duplicatus? I just know without YAML file you will get them

Reply:
Ahh sorry, bash script makes no sense, as clang-tidy requires access to all includes, therefore we need to run CMake to generate the moc and ui files. Thats why we have in our CI `CMAKE_GLOBAL_AUTOGEN_TARGET=ON` (requires CMake 3.14), so we don't need to build the whole project, instead we can run `make autogen` before. As the PCL has mostly legacy Qt code I don#t know how to trigger it, as autogen requires `AUTOUI`, `AUTORCC`, ...